accepted/0000-nohoist.md

```diff
@@ -0,0 +1,299 @@
+
+* Tradeoffs
+    
+    This is a classic simplicity vs usability. yarn can be simple and beautiful but if developers find it hard to use in practice, they will not use it. 
+
+# Alternatives
+
+* What other designs have been considered?
+
+    - There are a few [high level options](https://github.com/yarnpkg/yarn/issues/3882#issuecomment-338478889) discussed in [#3882](https://github.com/yarnpkg/yarn/issues/3882). 
+    - More implementation alternatives are also marked as discussion in their corresponding sections above.  
+    - we could choose not support nohoist-deep-dep(3), as use case (#2) is not very clear to me if we need this or what the expectation is. However, with inherited nohoistList, we get deep-nohoist pretty much for free so decided to just leave it in. We could either not document it publicly or leave it as it is and refine later when needed.
+
+# Unresolved questions
+
+## discussion-1
+**is allowing nohoist in public package bad?**
+
+As mentioned earlier, nohoist only impact yarn hoisting, for other package managers, it's a no-op and should be able to safely ignored... maybe I am missing some problematic use cases, here is what I got:
+
+Let's say a user project A has a dependency on a public package B, which has a dependency on 'react-native' that will need to be excluded from hoist otherwise it won't work. 
+
+||B has workspaces.nohoist=react-native|B didn't have workspaces.nohoist|
+|--|--|--|
+|A doesn't use yarn or workspace|no impact|no impact|
+|A uses yarn workspace|react-native will be excluded from hoisting without A doing anything|A needs to discover then put both B and react-native in its own workspaces.nohoist|
+|A uses yarn workspaces but excluded \_all\_|no impact, none will be hoisted|no impact, none will be hoisted|
+|A uses yarn workspaces but do not want to exclude react-native from hoisting (why?)|conflict! but the public package B is right and react-native should be excluded.|no conflict but B will most likely failed|
+
+In short, having unnecessary nohoist packages might cause inefficiency, but missing a necessary nohoist will lead to compile/execution error.
+
+## discussion-2
+**should the new hoisted package retain nohoist list?**
+example:
+```
+workspace-a -> A -> B
+```
+if developer added 'B' in the workspace-a's nohoist config:
+- package A will inherit nohist list ('B') from workspace-a
+- package A will be hoisted since it's not on the nohoist list ('B'), 
+- option-1: if we remove the nohoist list from A at this point, package B will be hoisted as well, the final graph will look like:
+```
+workspace-a
+A
+B
+```
+- option-2: if we don't remove the nohoist list from A, package B will stay under A as it is in the nohoist list:
+```
+workspace-a
+A -> B
+```
+IMHO, option-2 is more confusing than option-1. We can reason the nohoist inheritance is based on the dependency tree, once the tree is broken (by hoisting), so should the nohoist list inheritance. But I am open for better arguments.

```

Comment:
I understand that `nohoist` only works for workspace boundary, regular packages should not inherit workspaces' logic for hoisting.
So yeah, option-1 looks the right one

Reply:
@bestander 
> regular packages should not inherit workspaces' logic for hoisting

If by `regular packages` you mean the project without turning on `workspaces-experimental`, then yes nohoist only impacts the hoisting logic via `workspaces.nohoist` in package.json.
 
The more I think about the original question raised in the RFC, the more I realize that it has to be option-1. option-2 is not a valid alternative actually. Because a hoisted package can be "originated" from multiple sources that might have conflicting nohoist rules, therefore, inheritance-from-source is problematic. 

This also helps us simplify nohoist-inheritance concept, as it simply traverses the same node_module graph, just like package resolution. 

I will update the RFC to reflect this.